### ISSUE TITLE
BBC: Blead breaks Devel::NYTProf

### ISSUE THREAD

**Issue by jkeenan** [Perl/perl5#20490]
Here's a real mystery for you!

I have received two CPANtesters reports suggesting that a recent change in Perl blead has triggered test failures in CPAN distribution Devel-NYTProf (for which I am a co-maintainer).  When I bisect blead to identify the first bad commit, I get a reasonable result -- and I've now done that on two different OSes.  But when I build and install a `perl` at HEAD against which I install Devel::NYTProf, the tests PASS and the module is installed!

**The Failing CPANtesters reports**

http://www.cpantesters.org/cpan/report/73f36d72-5bfa-11ed-8236-4d2b72a47d1a
Linux / threaded build / gcc-10
Tested at 66964e141c6594291e27649d190bc6ca573861f2 (which was committed Thu Nov 3 06:52:48 2022)
`config_args='-des -Dprefix=~/bin/perl-blead -Dscriptdir=~/bin/perl-blead/bin -Dusedevel -Duse64bitall -Duseithreads'`

http://www.cpantesters.org/cpan/report/f6f0a66c-5cd7-11ed-84df-de89a64b8ed4
NetBSD / threaded build / gcc-5.5
Tested at 4cacf47cb8a44761f2f8360cb04a2a728b4ff9c8 (which was committed Fri Nov 4 19:49:48 2022 +0000)
`config_args='-des -Dprefix=~/bin/perl-blead -Dscriptdir=~/bin/perl-blead/bin -Dusedevel -Duse64bitall -Duseithreads'`

**The Test Failures**

The test failures are extensive; see http://www.cpantesters.org/cpan/report/73f36d72-5bfa-11ed-8236-4d2b72a47d1a.  Here are some of them:
```
perl5.37.6: NYTProf.xs:2416: subr_entry_setup: Assertion `sizeof(subr_entry_ix) == sizeof(SSize_t)' failed.
Aborted
Error status 34304 from "/home/cpan/bin/perl-blead/bin/perl5.37.6" -d:NYTProf -e "sub A { };" -e "1;" -e "A() "!
NYTPROF=calls=2:blocks=1:file=nytprof_readstream.out
Profile data incomplete, inflate error -5 ((null)) at end of input file, perhaps the process didn't exit cleanly or the file has been truncated  (refer to TROUBLESHOOTING in the documentation)
t/22-readstream.t ........ 
Dubious, test returned 134 (wstat 34304, 0x8600)
No subtests run 
t/30-util.t .............. ok
t/31-env.t ............... ok
perl5.37.6: NYTProf.xs:2416: subr_entry_setup: Assertion `sizeof(subr_entry_ix) == sizeof(SSize_t)' failed.
Aborted
Exit status 34304 from "/home/cpan/bin/perl-blead/bin/perl5.37.6" "-d:NYTProf" at t/40-savesrc.t line 23.

#   Failed test 'combination eval-ed okay'
#   at t/lib/NYTProfTest.pm line 182.
# Test group aborted: Profile data incomplete, inflate error -5 ((null)) at end of input file, perhaps the process didn't exit cleanly or the file has been truncated  (refer to TROUBLESHOOTING in the documentation)
# Looks like you planned 145 tests but ran 2.
# Looks like you failed 1 test of 2 run.
t/40-savesrc.t ........... 
Dubious, test returned 1 (wstat 256, 0x100)
Failed 144/145 subtests 
perl5.37.6: NYTProf.xs:2416: subr_entry_setup: Assertion `sizeof(subr_entry_ix) == sizeof(SSize_t)' failed.
Aborted
Exit status 34304 from "/home/cpan/bin/perl-blead/bin/perl5.37.6" "-d:NYTProf" at t/42-global.t line 27.

#   Failed test 'combination eval-ed okay'
#   at t/lib/NYTProfTest.pm line 182.
# Test group aborted: Profile data incomplete, inflate error -5 ((null)) at end of input file, perhaps the process didn't exit cleanly or the file has been truncated  (refer to TROUBLESHOOTING in the documentation)
# Looks like you planned 289 tests but ran 2.
# Looks like you failed 1 test of 2 run.
t/42-global.t ............ 
Dubious, test returned 1 (wstat 256, 0x100)
Failed 288/289 subtests 
perl5.37.6: NYTProf.xs:2416: subr_entry_setup: Assertion `sizeof(subr_entry_ix) == sizeof(SSize_t)' failed.
Aborted
Exit status 34304 from "/home/cpan/bin/perl-blead/bin/perl5.37.6" "-d:NYTProf" at t/44-model.t line 24.
Profile data incomplete, no terminator for 1 pids (refer to TROUBLESHOOTING in the NYTProf documentation)
Use of uninitialized value in division (/) at /home/cpan/.cpan/build/Devel-NYTProf-6.11-0/blib/lib/Devel/NYTProf/Data.pm line 112.
Use of uninitialized value in subtraction (-) at /home/cpan/.cpan/build/Devel-NYTProf-6.11-0/blib/lib/Devel/NYTProf/Data.pm line 113.
Can't resolve subinfo of 'UNIVERSAL::VERSION' at /home/cpan/.cpan/build/Devel-NYTProf-6.11-0/blib/lib/Devel/NYTProf/Data.pm line 339.
	Devel::NYTProf::Data::subinfo_of(Devel::NYTProf::Data=HASH(0x5564c4bdc500), "UNIVERSAL::VERSION") called at /home/cpan/.cpan/build/Devel-NYTProf-6.11-0/blib/lib/Devel/NYTProf/Data.pm line 705
	Devel::NYTProf::Data::file_line_range_of_sub(Devel::NYTProf::Data=HASH(0x5564c4bdc500), "UNIVERSAL::VERSION") called at t/44-model.t line 29
	main::__ANON__(undef, HASH(0x5564c57818d0)) called at t/lib/NYTProfTest.pm line 309
	NYTProfTest::__ANON__(HASH(0x5564c57818d0)) called at t/lib/NYTProfTest.pm line 182
	eval {...} called at t/lib/NYTProfTest.pm line 182
	NYTProfTest::do_foreach_opt_combination(ARRAY(0x5564c5781330), CODE(0x5564c5594db8)) called at t/lib/NYTProfTest.pm line 313
	NYTProfTest::run_test_group(HASH(0x5564c4ba98b0)) called at t/44-model.t line 33
Use of uninitialized value $first in concatenation (.) or string at t/44-model.t line 30.
Use of uninitialized value $last in concatenation (.) or string at t/44-model.t line 30.
```
I bisected twice: once with `clang10` in an unthreaded build on FreeBSD-12; once with `gcc-10` in a threaded build on Linux Debian 11.  So in the threaded build the invocation looked like this:
```
perl Porting/bisect.pl -Duseithreads \
--start=a589f4b93cf3864d7f665d0974e2d36f028b00b6 \
--end=f304415ff0355af37d5382ec491da9df928cc1be \
--module=Devel::NYTProf
```
Bisection in both cases pointed to this commit:
```
1a2b24d857d6903bb734c2fcb6cc2a72e1007059 is the first bad commit
commit 1a2b24d857d6903bb734c2fcb6cc2a72e1007059
Author: Tony Cook <tony@develop-help.com>
Date:   Thu Oct 20 15:04:51 2022 +1100
Commit:     Tony Cook <tony@develop-help.com>
CommitDate: Thu Nov 3 09:48:23 2022 +1100

    change the return value of SSNEW to SSize_t
```
(This CommitDate would be `Wed Nov 2 22:48:23 2022 +0000` in UTC, would it not?)

If you want to observe the bisection for yourself, this trimmed invocation will do it:
```
perl Porting/bisect.pl -Dusethreads \
--start=fce4124fa70860bf2675923721139e9b376b0856 \
--end=0ecd2f76f30a9cb78ed85d5e7805a221e27f643f \
--module=Devel::NYTProf
```
**Au Contraire!  The Module Installs**

Normally, with an extensive list of test failures due to a BBC commit, I would expect that:
* If I built a `perl` at a more recent commit than 1a2b24d857 and tried to install the CPAN distro in question, I would get test failures.
* Similarly, if I generated CPANtesters reports from such a `perl`, they would report FAILs.

But no, I have several times tbeen able to build perls at later commits and install Devel-NYTProf against them.

http://www.cpantesters.org/cpan/report/616f7076-5d33-11ed-aebd-643f06b3d53a
FreeBSD / threaded build / clang10
Tested at a0a15662a9734b83bd83642b9fecf9ba4a9b1a01 (CommitDate: Sat Nov 5 15:41:51 2022 +0000)

http://www.cpantesters.org/cpan/report/34bfacf2-5d39-11ed-8756-f3b0f8fb6917
Linux / threaded build / gcc-10
Tested at a0a15662a9734b83bd83642b9fecf9ba4a9b1a01 (CommitDate: Sat Nov 5 15:41:51 2022 +0000)

Even if I build and install at the "first bad commit" identified during bisection, I can install Devel::NYTProf:
```
$ ./bin/perl -v | head -2 | tail -1
This is perl 5, version 37, subversion 6 (v5.37.6 (v5.37.5-109-g1a2b24d857)) built for x86_64-linux-thread-multi
$ ./bin/perl -V:config_args
config_args='-des -Dusedevel -Dusethreads -Dprefix=/home/jkeenan/testing/1a2b24d857d6903bb734c2fcb6cc2a72e1007059 -Uversiononly -Dman1dir=none -Dman3dir=none';
$ ./bin/perl -Ilib -MDevel::NYTProf -E 'say $Devel::NYTProf::VERSION;'
6.11
fid 1 has no src saved for -e (NYTP_FIDf_HAS_SRC not set but src available!)
```
(That line beginning with `fid 1` is expected.)

Moreover, @cjguevara, the CPANtester who submitted these reports, has re-run the process on each of the two platforms on which failures were originally observed -- and those reports are now PASSing.

http://www.cpantesters.org/cpan/report/0eb85b9c-5c7d-11ed-b923-fda63538bed4
Linux / threaded build / gcc-10
Tested at 66964e141c6594291e27649d190bc6ca573861f2 (CommitDate: Thu Nov 3 10:52:48 2022)

http://www.cpantesters.org/cpan/report/dc2cbd32-5d74-11ed-a351-8d70432d9cfe
NetBSD / threaded build / gcc-5.5
Tested at 4cacf47cb8a44761f2f8360cb04a2a728b4ff9c8 (CommitDate: Fri Nov 4 19:49:48 2022)

**Possibly Relevant Source Code**

NYTProf.xs (from [version 6.11 release tarball](https://cpan.metacpan.org/authors/id/J/JK/JKEENAN/Devel-NYTProf-6.11.tar.gz)
```
2393 static I32
2394 subr_entry_setup(pTHX_ COP *prev_cop, subr_entry_t *clone_subr_entry, OPCODE op_type, SV *subr_sv     )
2395 {
2396     int saved_errno = errno;
2397     subr_entry_t *subr_entry;
2398     I32 prev_subr_entry_ix;
2399     subr_entry_t *caller_subr_entry;
2400     const char *found_caller_by;
2401     char *file;
2402 
2403     /* allocate struct to save stack (very efficient) */
2404     /* XXX "warning: cast from pointer to integer of different size" with use64bitall=define */
2405     prev_subr_entry_ix = subr_entry_ix;
2406     subr_entry_ix = SSNEWa(sizeof(*subr_entry), MEM_ALIGNBYTES);
2407 
2408     if (subr_entry_ix <= prev_subr_entry_ix) {
2409         /* one cause of this is running NYTProf with threads */
2410         logwarn("NYTProf panic: stack is confused, giving up! (Try running with subs=0) ix=%" IVd     f " prev_ix=%" IVdf "\n", (IV)subr_entry_ix, (IV)prev_subr_entry_ix);
2411         /* limit the damage */
2412         disable_profile(aTHX);
2413         return prev_subr_entry_ix;
2414     }
2415 
2416     subr_entry = subr_entry_ix_ptr(subr_entry_ix);  /* <-- line cited in bisection */
2417     Zero(subr_entry, 1, subr_entry_t);
2418 
```
To sum up, we have some CPANtester failure reports that can be bisected to a particular, very recent commit.  But when we build at that commit or later, we don't get failures in the affected CPAN distro.

Anyone have any ideas?

@tonycoz @timbunce

**Comment by jkeenan** [comment]
An additional CPANtesters failure:
http://www.cpantesters.org/cpan/report/968ae754-5dce-11ed-895b-601f4db3123d
FreeBSD-13 / unthreaded build / clang13
Tested at b65a15b678acb393729f0f74c0bfd167f5dac941 (committed Sat Nov 05)

**Comment by tonycoz** [comment]
Only debugging builds are failing, which is reasonable, since it's failing on an assertion.

I've opened https://github.com/timbunce/devel-nytprof/pull/200

**Comment by jkeenan** [comment]
> Only debugging builds are failing, which is reasonable, since it's failing on an assertion.
> 

Does that mean that when I call `Porting/bisect.pl`, I'm requesting a `-DDEBUGGING` build?  If so, that's news to me!

> I've opened [timbunce/devel-nytprof#200](https://github.com/timbunce/devel-nytprof/pull/200)

Am testing your p.r. as best I can.

**Comment by jkeenan** [comment]
> Only debugging builds are failing, which is reasonable, since it's failing on an assertion.
> 

@tonycoz, that is not the case.  See http://fast-matrix.cpantesters.org/?dist=Devel-NYTProf;perl=5.37.6;reports=1#sl=7,1, where none of the reports (PASS or FAIL) have been run on debugging builds.  Example:
http://www.cpantesters.org/cpan/report/968ae754-5dce-11ed-895b-601f4db3123d

**Comment by tonycoz** [comment]
Actually, it wasn't -DDEBUGGING, which I assumed because of the assertions.

Devel::NYTProf [enables](https://github.com/timbunce/devel-nytprof/blob/master/NYTProf.xs#L45) [assertions ](https://github.com/timbunce/devel-nytprof/blob/master/Makefile.PL#L52)for non-DEBUGGING builds when AUTOMATED_TESTING is set, which is true for CPAN testers, and bisect-runner.pl (which does the heavy lifting for bisect.pl) also sets AUTOMATED_TESTING.

**Comment by AnFunctionArray** [comment]
@tonycoz It's the assertions you have added on https://github.com/Perl/perl5/commit/1a2b24d857d6903bb734c2fcb6cc2a72e1007059#diff-80f18c4657ae7b33549e6da73aa781cd3959a14eb19982200acd92bad261a0c5R288

I believe

**Comment by tonycoz** [comment]
> @tonycoz It's the assertions you have added on [1a2b24d#diff-80f18c4657ae7b33549e6da73aa781cd3959a14eb19982200acd92bad261a0c5R288](https://github.com/Perl/perl5/commit/1a2b24d857d6903bb734c2fcb6cc2a72e1007059#diff-80f18c4657ae7b33549e6da73aa781cd3959a14eb19982200acd92bad261a0c5R288)

Exactly right.  I added the assertions to illustrate code broken by the change as early as possible.  Devel::NYTProf is the only CPAN module I could find using SSNEW*()

**Comment by AnFunctionArray** [comment]
@tonycoz Ok I see.

**Comment by jkeenan** [comment]
> Actually, it wasn't -DDEBUGGING, which I assumed because of the assertions.
> 
> Devel::NYTProf [enables](https://github.com/timbunce/devel-nytprof/blob/master/NYTProf.xs#L45) [assertions ](https://github.com/timbunce/devel-nytprof/blob/master/Makefile.PL#L52)for non-DEBUGGING builds when AUTOMATED_TESTING is set, which is true for CPAN testers, and bisect-runner.pl (which does the heavy lifting for bisect.pl) also sets AUTOMATED_TESTING.

Aha!  That would explain the difference between, *e.g.,* these two CPANtesters reports which were apparently run on the same machine, the first a FAIL, the second a PASS.

FAIL: http://www.cpantesters.org/cpan/report/968ae754-5dce-11ed-895b-601f4db3123d
PASS: http://www.cpantesters.org/cpan/report/d905689c-5e1a-11ed-80aa-3f98ab092510

The FAIL has `AUTOMATED_TESTING = 1` and `TERM = screen`.

The PASS does not define `AUTOMATED_TESTING` and has `TERM = xterm`.

This suggests that the tester (@cjg-cguevara) spotted the FAIL, then went to a terminal and manually generated a new CPANtesters report which, because the environment did not have `AUTOMATED_TESTING` defined, was graded as PASS.

**Comment by jkeenan** [comment]
I've released Devel-NYTProf-6.12 to CPAN incorporating a patch from @tonycoz.  I am monitoring the [CPANtesters reports](http://fast-matrix.cpantesters.org/?dist=Devel-NYTProf) and will close this ticket after our Nov 20 development release.
